llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -1428,6 +1428,40 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
   return nullptr;
 }
 
+/// Fold an unsigned minimum of trailing or leading zero bits counts:
+///   umin(cttz(CtOp, ZeroUndef), ConstOp) --> cttz(CtOp | (1 << ConstOp))
+///   umin(ctlz(CtOp, ZeroUndef), ConstOp) --> ctlz(CtOp | ((1 << (bitwidth-1))
+///                                              >> ConstOp))
+template <Intrinsic::ID IntrID>
+static Instruction *foldMinimumOverTrailingOrLeadingZeroCount(
+    Instruction *OrigInst, Value *CtOp, Value *ZeroUndef, Constant *ConstOp,
+    const DataLayout &DL, InstCombiner::BuilderTy &Builder) {
+  static_assert(IntrID == Intrinsic::cttz || IntrID == Intrinsic::ctlz,
+                "This helper only supports cttz and ctlz intrinsics");
+
+  auto BitWidth = ConstOp->getType()->getScalarSizeInBits();
+  auto *Ty = ConstOp->getType();
+
+  Constant *NewConst = ConstantFoldSelectInstruction(
+      ConstantFoldCompareInstOperands(CmpInst::ICMP_ULT, ConstOp,
+                                      ConstantInt::get(Ty, BitWidth), DL),
+      ConstantFoldBinaryOpOperands(
+          IntrID == Intrinsic::cttz ? Instruction::Shl : Instruction::LShr,
+          IntrID == Intrinsic::cttz
+              ? ConstantInt::get(Ty, 1)
+              : ConstantInt::get(Ty, APInt::getSignedMinValue(BitWidth)),
+          ConstOp, DL),
+      Constant::getNullValue(Ty));
```

Comment:
If you go with this implementation, you'll also have to adjust the alive2 proofs in the PR description to be in terms of this implementation rather than assumes.

Reply:
Does this still hold for the new implementation? The branches are now explicitly stated in the code, which is why the proofs still rely on assumes.